File: 백준/Silver/12789. 도키도키 간식드리미/도키도키 간식드리미.cc
#include <iostream>
#include <stack>
#include <vector>

int main() {
    int n;
    std::cin >> n;

    std::vector<int> line(n);
    for (int i = 0; i < n; ++i)
    {
        std::cin >> line[i];
    }

    std::stack<int> side;
    int next = 1;
    for (int i = 0; i < n; ++i) 
    {
        const int& current = line[i];
        while (!side.empty() && side.top() == next) 
        {
            side.pop();
            next++;
        }
        if (current == next) 
        {
            next++;
        }
        else 
        {
            side.push(current);
        }
    }

    while (!side.empty() && side.top() == next) 
    {
        side.pop();
        next++;
    }

    if (side.empty()) 
    {
        std::cout << "Nice";
    }
    else 
    {
        std::cout << "Sad";
    }

    return 0;
}

File: 백준/Silver/12789. 도키도키 간식드리미/도키도키 간식드리미.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.StringTokenizer;

public class Main {

	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		int n = Integer.parseInt(br.readLine());
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		ArrayDeque<Integer> stack = new ArrayDeque<Integer>();
		int target = 1;
		for(int i = 0; i < n; i++) {
			int num = Integer.parseInt(st.nextToken());
			if(num == target) {
				target++;
				while(!stack.isEmpty()) {
					int top = stack.getFirst();
					if(top == target) {
						stack.pop();
						target++;
					}else {
						break;
					}
				}
			}else {
				stack.push(num);
			}
		}
		String answer = stack.isEmpty() ? "Nice" : "Sad";
		System.out.println(answer);
	}

}
